### PR TITLE
BUG: Attributes skipped when serialising plain Python objects to JSON (#42768)

### LINKED ISSUE #42768: BUG: to_json() swallows attributes of dataclasses
#### Code Sample, a copy-pastable example

```python
from dataclasses import dataclass
import pandas as pd

@dataclass()
class MyObject:
    a: "str"
    b: "str"
    c: "str"

@dataclass()
class MyObject2:
    a: "str"
    _b: "str"
    c: "str"
        

o1 = MyObject(a='a', b='b', c='c')
o2 = MyObject2(a='a', _b='b', c='c')


df = pd.DataFrame({"objects":[o1, o2]})

print(df.to_json())
print(pd._libs.json.dumps(o1))
print(pd._libs.json.dumps(o2))
```

Output:
```
{"objects":{"0":{"a":"a","c":"c"},"1":{"a":"a"}}}
{"a":"a","c":"c"}
{"a":"a"}
```

#### Problem description

While the dataclasses have three attributes depending on the naming only one or two show up in the JSON dump.

#### Expected Output

```
{"objects":{"0":{"a":"a","b"="b","c":"c"},"1":{"a":"a","_b"="b","c":"c"}}}
{"a":"a","b"="b","c":"c"}
{"a":"a","_b"="b","c":"c"}
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : c7f7443c1bad8262358114d5e88cd9c8a308e8aa
python           : 3.9.1.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.5.0
Version          : Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.UTF-8

pandas           : 1.3.1
numpy            : 1.21.1
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.0.1
setuptools       : 49.2.1
Cython           : None
pytest           : 6.2.4
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.9.1 (dt dec pq3 ext lo64)
jinja2           : 2.11.3
IPython          : 7.20.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 2021.07.0
fastparquet      : None
gcsfs            : 2021.07.0
matplotlib       : 3.3.4
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : 0.15.0
pyarrow          : 4.0.1
pyxlsb           : None
s3fs             : None
scipy            : 1.7.0
sqlalchemy       : 1.4.22
tables           : None
tabulate         : 0.8.9
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : 0.51.2
</details>

### PR DESCRIPTION
**Pull request by joelgibson**
This patch closes #42768 and fixes a previously-reported issue #33043.

When `ujson` falls back to encoding a Python object, it iterates over all of the non-callable attributes of the object which do not start with an underscore, using them as JSON keys. While iterating the `dirs(obj)` list, an index was incremented twice causing every second attribute to be skipped.

I've fixed this increment, and moved some of the declarations inside the function so that control flow is more clear. I've also added a regression test for this behaviour.

- [x] closes #42768
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

This patch also xref #41174, but I'm not sure that turning complex numbers into `{"real": ..., "imag": ...}` JSON objects was intended in the first place, so I haven't added a test for this.

### CODE REVIEW

File: doc/source/whatsnew/v1.3.2.rst [hunk 1]
```diff
@@ -36,6 +36,7 @@ Bug fixes
 - :meth:`.Styler.hide_columns` now hides the index name header row as well as column headers (:issue:`42101`)
 - :meth:`.Styler.set_sticky` has amended CSS to control the column/index names and ensure the correct sticky positions (:issue:`42537`)
 - Bug in de-serializing datetime indexes in PYTHONOPTIMIZED mode (:issue:`42866`)
+- Some attributes were skipped when serialising plain Python objects to JSON (:issue:`42768`, :issue:`33043`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.2.rst:39]: move to 1.4

**Review by jreback — CHANGES_REQUESTED** [review]
does this have any user visible effects in to_json itself?

**Comment by jreback** [comment]
can you replicate the test in the OP

**Comment by joelgibson** [comment]
@jreback The test more or less duplicates the OP, just using regular classes rather than dataclasses. However, the expected output in the OP included properties beginning with underscores, wheras I have left out of the resulting JSON because this is what the existing (intended, I assume) behaviour did. 

It does have user-visible effects, for example if classes are stored as elements in series or dataframes. The program:

```
import pandas as pd

class A:
    def __init__(self, a, b):
        self.a = a
        self.b = b

series = pd.Series([A(a=1, b=2)])
print(series.to_json())
```

used to return `{"0":{"a":1}}`, and now returns `{"0":{"a":1,"b":2}}`.

File: doc/source/whatsnew/v1.4.0.rst [hunk 2]
```diff
@@ -258,6 +258,7 @@ I/O
 - Bug in :func:`read_excel` attempting to read chart sheets from .xlsx files (:issue:`41448`)
 - Bug in :func:`json_normalize` where ``errors=ignore`` could fail to ignore missing values of ``meta`` when ``record_path`` has a length greater than one (:issue:`41876`)
 - Bug in :func:`read_csv` with multi-header input and arguments referencing column names as tuples (:issue:`42446`)
+- Some attributes were skipped when serialising plain Python objects to JSON (:issue:`42768`, :issue:`33043`)
```
Comment (mzeitlin11) [inline doc/source/whatsnew/v1.4.0.rst:261]: To give the user-facing effect, can we link to `to_json` here?

File: pandas/tests/io/json/test_ujson.py [hunk 3]
```diff
@@ -720,6 +720,21 @@ def my_obj_handler(_):
...
+                self.a = a
+                self.b = b
+                self._c = _c
+                self.d = d
+
+            def e(self):
+                return 5
+
+        test_object = _TestObject(a=1, b=2, _c=3, d=4)
+        assert ujson.decode(ujson.encode(test_object)) == {"a": 1, "b": 2, "d": 4}
```
Comment (mzeitlin11) [inline pandas/tests/io/json/test_ujson.py:736]: This is a great test - can we add an analogous test that hits `to_json` to test the user-facing effect?

**Review by mzeitlin11 — COMMENTED** [review]
Thanks for the pr @joelgibson! While I agree that the changes you made make the code more readable, would it possible to leave those for a followup pr? Just to make the changes that actually affect code logic clearer

File: pandas/tests/io/json/test_ujson.py (hunk 3 above)
Comment (joelgibson) [reply pandas/tests/io/json/test_ujson.py:736]: Sure, done.

File: pandas/tests/io/json/test_pandas.py [hunk 4]
```diff
@@ -1769,3 +1769,18 @@ def test_to_json_multiindex_escape(self):
             "\"(Timestamp('2017-01-23 00:00:00'), 'bar')\":true}"
         )
         assert result == expected
+
+    def to_json_series_of_objects(self):
+        class _TestObject:
```
Comment (jreback) [inline pandas/tests/io/json/test_pandas.py:1774]: can you add the issue reference as a comment on this tests and the one below.

File: doc/source/whatsnew/v1.4.0.rst [hunk 5]
```diff
@@ -258,6 +258,7 @@ I/O
 - Bug in :func:`read_excel` attempting to read chart sheets from .xlsx files (:issue:`41448`)
 - Bug in :func:`json_normalize` where ``errors=ignore`` could fail to ignore missing values of ``meta`` when ``record_path`` has a length greater than one (:issue:`41876`)
 - Bug in :func:`read_csv` with multi-header input and arguments referencing column names as tuples (:issue:`42446`)
+- Bug in :func:`to_json` where some attributes were skipped when serialising plain Python objects to JSON (:issue:`42768`, :issue:`33043`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:261]: should be Series.to_json

**Review by jreback — CHANGES_REQUESTED** [review]
some test comments, ping on green.

File: pandas/tests/io/json/test_pandas.py (hunk 4 above)
Comment (joelgibson) [reply pandas/tests/io/json/test_pandas.py:1774]: The comment is slightly further down the test so it's closer to the expected output, is that ok?

File: doc/source/whatsnew/v1.4.0.rst (hunk 5 above)
Comment (joelgibson) [reply doc/source/whatsnew/v1.4.0.rst]: I've added `Series.to_json` and `Dataframe.to_json`.

**Review by jreback — APPROVED** [review]

**Comment by joelgibson** [comment]
Thanks @mzeitlin11 - I would usually only change the part of the function which affected the logic, but the logic of that function was so unclear it seemed to warrant a cleanup anyway. For example on 928 it looks like something interesting is being assigned but actually NULL is being assigned. (Incidentally, how should I go about putting in the cleanup PR - just title it with "CLN" and reference this discussion?). 

@jreback ping

File: pandas/tests/io/json/test_pandas.py [hunk 6]
```diff
@@ -1769,3 +1769,18 @@ def test_to_json_multiindex_escape(self):
             "\"(Timestamp('2017-01-23 00:00:00'), 'bar')\":true}"
         )
         assert result == expected
+
+    def to_json_series_of_objects(self):
```
Comment (mzeitlin11) [inline pandas/tests/io/json/test_pandas.py:1773]:
```suggestion
    def test_to_json_series_of_objects(self):
```
Having `test` in the name is necessary for the test to be discovered and run during CI

**Review by mzeitlin11 — COMMENTED** [review]
LGTM, thanks @joelgibson! Followup with your original cleanups would be great!

> (Incidentally, how should I go about putting in the cleanup PR - just title it with "CLN" and reference this discussion?)

Yep sounds good, nothing special needs to be done

**Comment by jreback** [comment]
thanks @joelgibson

**Pull request merged**